modules/tracking/src/tldTracker.cpp

```diff
@@ -0,0 +1,297 @@
+ // or tort (including negligence or otherwise) arising in any way out of
+ // the use of this software, even if advised of the possibility of such damage.
+ //
+ //M*/
+
+#include "tldTracker.hpp"
+
+
+namespace cv
+{
+
+Ptr<TrackerTLD> TrackerTLD::createTracker(const TrackerTLD::Params &parameters)
+{
+    return Ptr<tld::TrackerTLDImpl>(new tld::TrackerTLDImpl(parameters));
+}
+
+namespace tld
+{
+
+TrackerTLDImpl::TrackerTLDImpl(const TrackerTLD::Params &parameters) :
+    params( parameters )
+{
+  isInit = false;
+  trackerProxy = Ptr<TrackerProxyImpl<TrackerMedianFlow, TrackerMedianFlow::Params> >
+      (new TrackerProxyImpl<TrackerMedianFlow, TrackerMedianFlow::Params>());
+}
+
+void TrackerTLDImpl::read(const cv::FileNode& fn)
+{
+  params.read( fn );
+}
+
+void TrackerTLDImpl::write(cv::FileStorage& fs) const
+{
+  params.write( fs );
+}
+
+bool TrackerTLDImpl::initImpl(const Mat& image, const Rect2d& boundingBox)
+{
+    Mat image_gray;
+    trackerProxy->init(image, boundingBox);
+    cvtColor( image, image_gray, COLOR_BGR2GRAY );
+    data = Ptr<Data>(new Data(boundingBox));
+    double scale = data->getScale();
+    Rect2d myBoundingBox = boundingBox;
+    if( scale > 1.0 )
+    {
+        Mat image_proxy;
+        resize(image_gray, image_proxy, Size(cvRound(image.cols * scale), cvRound(image.rows * scale)), 0, 0, DOWNSCALE_MODE);
+        image_proxy.copyTo(image_gray);
+        myBoundingBox.x *= scale;
+        myBoundingBox.y *= scale;
+        myBoundingBox.width *= scale;
+        myBoundingBox.height *= scale;
+    }
+    model = Ptr<TrackerTLDModel>(new TrackerTLDModel(params, image_gray, myBoundingBox, data->getMinSize()));
+
+    data->confident = false;
+    data->failedLastTime = false;
+
+    return true;
+}
+
+bool TrackerTLDImpl::updateImpl(const Mat& image, Rect2d& boundingBox)
+{
+    Mat image_gray, image_blurred, imageForDetector;
+    cvtColor( image, image_gray, COLOR_BGR2GRAY );
+    double scale = data->getScale();
+    if( scale > 1.0 )
+        resize(image_gray, imageForDetector, Size(cvRound(image.cols*scale), cvRound(image.rows*scale)), 0, 0, DOWNSCALE_MODE);
+    else
+        imageForDetector = image_gray;
+    GaussianBlur(imageForDetector, image_blurred, GaussBlurKernelSize, 0.0);
+    TrackerTLDModel* tldModel = ((TrackerTLDModel*)static_cast<TrackerModel*>(model));
+    data->frameNum++;
+    Mat_<uchar> standardPatch(STANDARD_PATCH_SIZE, STANDARD_PATCH_SIZE);
+    std::vector<TLDDetector::LabeledPatch> detectorResults;
+    //best overlap around 92%
+
+    std::vector<Rect2d> candidates;
+    std::vector<double> candidatesRes;
+    bool trackerNeedsReInit = false;
+    for( int i = 0; i < 2; i++ )
+    {
+        Rect2d tmpCandid = boundingBox;
+        if( ( (i == 0) && !data->failedLastTime && trackerProxy->update(image, tmpCandid) ) ||
+			((i == 1) && (tldModel->detector->ocl_detect(imageForDetector, image_blurred, tmpCandid, detectorResults, tldModel->getMinSize()))))
```

Comment:
Maybe here a condition if is needed for ocl_detect and detect methods, see this presentation https://docs.google.com/presentation/d/1qoa29N_B-s297-fp0-b3rBirvpzJQp8dCtllLQ4DVCY/present#slide=id.p26